server/app/jobs/grid_service_scheduler_worker.rb

```diff
@@ -9,58 +9,63 @@ def initialize(autostart = true)
 
   def watch
     loop do
-      self.check_deploy_queue
+      if service_deploy = self.check_deploy_queue
+        self.deploy(service_deploy)
+      end
       sleep 1
     end
   end
 
+  # Fetch deploy from queue, and return it if it should be run
+  #
+  # @return [GridServiceDeploy, nil] deploy to run
   def check_deploy_queue
     service_deploy = fetch_deploy_item
-    return unless service_deploy
+    return nil unless service_deploy
 
     with_dlock("check_deploy_queue:#{service_deploy.grid_service_id}", 10) do
-      service_deploy.reload
-      if service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
-        self.perform(service_deploy)
+      if service_deploy.grid_service.deploy_running?
+        info "delaying #{service_deploy.grid_service.to_path} deploy because there is another deploy in progress"
+        return nil
+
+      elsif service_deploy.grid_service.running? || service_deploy.grid_service.initialized?
+        info "starting #{service_deploy.grid_service.to_path} deploy"
+        service_deploy.set(started_at: Time.now.utc)
+        return service_deploy
+
       else
-        service_deploy.destroy
+        info "aborting #{service_deploy.grid_service.to_path} deploy of non-running service"
+        service_deploy.abort! "service is not running"
+        return nil
       end
     end
   end
 
+  # Pick up the oldest non-queued deploy, mark it as queued, and return for processing.
+  # The caller has 30s to process the returned deploy, or it can be picked up again.
+  #
+  # XXX: this can pick up an aborted deploy if it not started_at?
+  #
   # @return [GridServiceDeploy, NilClass]
   def fetch_deploy_item
-    GridServiceDeploy.where(started_at: nil)
+    GridServiceDeploy.any_of({:queued_at => nil}, {:queued_at.lt => 30.seconds.ago, :started_at => nil})
```

Comment:
AFAIK this works as intended, it's a single moped command using indexed fields:

```
DEBUG -- :   MOPED: 127.0.0.1:27017 COMMAND      database=kontena_development command={:findAndModify=>"grid_service_deploys", :query=>{"$or"=>[{"queued_at"=>nil}, {"queued_at"=>{"$lt"=>2017-04-27 15:16:07 UTC}, "started_at"=>nil}]}, :new=>true, :sort=>{"created_at"=>1}, :update=>{:$set=>{:queued_at=>2017-04-27 15:16:37 UTC}}} runtime: 1.3541ms
```

The significant design choice here is that it's safe against worker failure after having dequeued the deploy... the deploy will naturally be re-dequeued in time.